[jit] add `hasattr()`

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29333 [jit] move some tests to test_builtins.py
* **#29332 [jit] add `hasattr()`**

Even though we're statically typed, this can be useful, e.g. as
shorthand when iterating through a module list.

Theoretically this could be dynamic--unlike `getattr` or something we
don't need static information about the attribute's type. It's more
general that way, but less consistent with our other stuff. If people
think it's better as a dynamic check then let me know.

Differential Revision: [D18393097](https://our.internmc.facebook.com/intern/diff/D18393097)

suo committed:
[jit] add `hasattr()`

Even though we're statically typed, this can be useful, e.g. as
shorthand when iterating through a module list.

Theoretically this could be dynamic--unlike `getattr` or something we
don't need static information about the attribute's type. It's more
general that way, but less consistent with our other stuff. If people
think it's better as a dynamic check then let me know.

suo committed:
Update on "[jit] add `hasattr()`"


Even though we're statically typed, this can be useful, e.g. as
shorthand when iterating through a module list.

Theoretically this could be dynamic--unlike `getattr` or something we
don't need static information about the attribute's type. It's more
general that way, but less consistent with our other stuff. If people
think it's better as a dynamic check then let me know.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -2379,6 +2380,25 @@ struct to_ir {
+          return std::make_shared<SimpleValue>(graph->insertConstant(hasAttr));
+        } else {
+          throw ErrorReport(apply)
+              << "hasattr's first argument must be a class type, got "
```
"must be an object", must be a class type means it is `hasattr(MyClass, "foo")`

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -2379,6 +2380,25 @@ struct to_ir {
       }
+      case prim::HasAttr: {
+        checkApplyNumInputs(apply, 2);
+        auto obj = emitSugaredExpr(apply.inputs()[0], 1);
```
just call emitExpr if asValue will be called later.

zdevito reviewed (approved):
Looks good, but how can it be used? You get a true/false constant in the body, but then you will still get an error if the program statically contains a getattr call. Consider making it a case of emitCondValue where it can cause if statements to get specialized. We already do it for attributes having a None type vs. a different type.

suo commented:
Hm good point—the motivating example didn't include a follow-up call to `getattr`, but it is obviously something that someone would expect to work. I'll do the `emitCondValue` thing.

kostmo commented:
## CircleCI build failures summary
As of commit 519aff2:
* 0/7 recognized as flaky
* 7/7 broken upstream. You may want to rebase on the [latest viable branch](https://dr.pytorch.org/master-viable-commits.html).
* 0/7 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=519aff2a4b35061c348706945ef4564a6ace3d11).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

suo committed:
Update on "[jit] add `hasattr()`"


Even though we're statically typed, this can be useful, e.g. as
shorthand when iterating through a module list.

Theoretically this could be dynamic--unlike `getattr` or something we
don't need static information about the attribute's type. It's more
general that way, but less consistent with our other stuff. If people
think it's better as a dynamic check then let me know.

suo committed:
Update on "[jit] add `hasattr()`"


Even though we're statically typed, this can be useful, e.g. as
shorthand when iterating through a module list.

Theoretically this could be dynamic--unlike `getattr` or something we
don't need static information about the attribute's type. It's more
general that way, but less consistent with our other stuff. If people
think it's better as a dynamic check then let me know.

Differential Revision: [D18393097](https://our.internmc.facebook.com/intern/diff/D18393097)
